### PR TITLE
[SPARK-40511][BUILD][CORE] Upgrade slf4j to 2.0.3

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims to upgrade slf4j from 1.7.36 to 2.0.3, the main change as follows:

1. Upgrade slf4j version from 1.7.36 to 2.0.3
2. Use ` log4j-slf4j2-impl` instead of ` log4j-slf4j-impl` refer to https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/index.html
3. Adapt to the implementation of `Logging.isLog4j2()`, slf4j2 uses `SLF4JServiceProvider` instead of `StaticLoggerBinder`

### Why are the changes needed?
slf4j 1.7.36 is no longer actively developed, and the current stable and actively developed version of slf4j is 2.0.3.

- https://www.slf4j.org/news.html
- https://www.slf4j.org/faq.html#changesInVersion200
- https://www.slf4j.org/download.html

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

**Comment by LuciferYang** [comment]
Test API compatibility first

File: pom.xml [hunk 1]
```diff
@@ -112,7 +112,7 @@
     <maven.version>3.8.6</maven.version>
     <exec-maven-plugin.version>1.6.0</exec-maven-plugin.version>
     <sbt.project.name>spark</sbt.project.name>
-    <slf4j.version>1.7.36</slf4j.version>
+    <slf4j.version>2.0.0</slf4j.version>
     <log4j.version>2.18.0</log4j.version>
```
Comment (martin-g) [inline pom.xml:116]: Log4j 2.18.0 does not support SLF4J 2.x. See https://github.com/apache/logging-log4j2/commits/968371c7a88ceb1c1ce15b70e3a5a08b49be7169/log4j-slf4j20-impl/pom.xml?browsing_rename_history=true&new_path=log4j-slf4j2-impl/pom.xml&original_branch=release-2.x
I think you will have to wait for LOG4J 2.19.

Slf4j 1.x uses static binding, whilst SLF4J 2.x uses ServiceLoader.

Comment (LuciferYang) [reply pom.xml:116]: Thanks @martin-g ~

**Comment by LuciferYang** [comment]
close and wait log4j 2.19.x

**Comment by martin-g** [comment]
@LuciferYang https://search.maven.org/artifact/org.apache.logging.log4j/log4j-slf4j2-impl has been released!

**Comment by LuciferYang** [comment]
Thanks @martin-g

**Comment by LuciferYang** [comment]
reopen this due to log4j 2.19.0 released and Spark already upgrade to use this version, test first

File: common/kvstore/pom.xml [hunk 2]
```diff
@@ -89,7 +89,7 @@
     </dependency>
     <dependency>
       <groupId>org.apache.logging.log4j</groupId>
-      <artifactId>log4j-slf4j-impl</artifactId>
+      <artifactId>log4j-slf4j2-impl</artifactId>
```
Comment (LuciferYang) [inline common/kvstore/pom.xml:92]: https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/index.html

It seems that `log4j-slf4j-impl` and `log4j-slf4j2-impl` should not be used together

File: pom.xml [hunk 3]
```diff
@@ -112,7 +112,7 @@
     <maven.version>3.8.6</maven.version>
     <exec-maven-plugin.version>1.6.0</exec-maven-plugin.version>
     <sbt.project.name>spark</sbt.project.name>
-    <slf4j.version>1.7.36</slf4j.version>
+    <slf4j.version>2.0.0</slf4j.version>
```
Comment (martin-g) [inline pom.xml:115]: 2.0.1 has been released a week ago

Comment (LuciferYang) [reply pom.xml:115]: Some code changes are needed, let me fix them first

File: core/src/main/scala/org/apache/spark/internal/Logging.scala [hunk 4]
```diff
@@ -226,11 +226,18 @@ private[spark] object Logging {
   }
 
   private def isLog4j2(): Boolean = {
+    import java.util.ServiceLoader
     // This distinguishes the log4j 1.2 binding, currently
     // org.slf4j.impl.Log4jLoggerFactory, from the log4j 2.0 binding, currently
     // org.apache.logging.slf4j.Log4jLoggerFactory
-    val binderClass = StaticLoggerBinder.getSingleton.getLoggerFactoryClassStr
-    "org.apache.logging.slf4j.Log4jLoggerFactory".equals(binderClass)
+    val loader = Utils.getContextOrSparkClassLoader
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/internal/Logging.scala:233]: maybe incorrect

**Comment by LuciferYang** [comment]
friendly ping @viirya @HyukjinKwon @dongjoon-hyun @srowen 

This pr try to upgrade slf4j from 1.x to 2.x, but I don't think it is an appropriate time to upgrade it now due to slf4j 2.0.x still released frequently, so I will keep this pr as draft for a period of time and continuous testing it.

**Review by martin-g — APPROVED** [review]
LGTM!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @LuciferYang and @martin-g .

cc @viirya

**Comment by LuciferYang** [comment]
> > so I will keep this pr as draft for a period of time and continuous testing it.
> 
> So this is changed?

may be a mistake, re-change to draft

**Comment by LuciferYang** [comment]
mvn test `Java8 + hadoop-2 profile` with this pr , all test passed

**Comment by LuciferYang** [comment]
2.0.3 released, let us wait more time

**Comment by LuciferYang** [comment]
mvn test `Java17 + hadoop-3` profile with this pr , all test passed

**Review by HyukjinKwon — APPROVED** [review]

**Comment by srowen** [comment]
Is this ready to merge?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by viirya — APPROVED** [review]

**Review by itholic — APPROVED** [review]
LGTM

**Comment by LuciferYang** [comment]
should we merge this one first ?

**Comment by dongjoon-hyun** [comment]
Yep. Thank you, @LuciferYang and all.
Merged to master.

**Comment by LuciferYang** [comment]
Thanks @martin-g @dongjoon-hyun @HyukjinKwon @srowen @viirya @itholic

**Pull request closed without merging**